### ISSUE TITLE
Cargo.nix Why does it need to live here?

### ISSUE THREAD

**Issue by krisnova** [aurae-runtime/aurae#398]
I understand that `Cargo.nix` is a file that "pins" the Cargo dependencies to a specific version/release/etc when compiling Aurae for Nix.

My question isn't what the file it **is** or what the file **does**... my question is why is it in our repository? 

## Few Questions

 - Is this file something that the project should support? 
 - Do we intend on supporting Nix compiling Aurae from scratch in favor of downloading a signed artifact? 
 - Do we intend to keep these operating system specific files in our core repo? 
 - Should these operating system specific configs be moved to other repos? (For example: a nix repo?)

---

Basically my concern is that I am not sure why the project needs to track 4,500+ lines of `.nix` data in our core repo? It is polluting our metrics and bloating the repository. 

![image](https://user-images.githubusercontent.com/13757818/218329718-4ece0563-833f-4578-be05-b80e11b06727.png)

**Comment by hbjydev** [comment]
Just one question to answer for me :) I don't wanna get involved in "should it be here/do we wanna care" discussions because I've made exactly 1 PR lol.

> Do we intend on supporting Nix compiling Aurae from scratch in favor of downloading a signed artifact?

This just _won't work_ if you run NixOS, as Nix poisons linking for stuff it builds in a way that you can only run it with Nix. However, _NixOS_ (the operating system) relies _totally_ on this and doesn't follow the Linux FHS for the most part, so `/lib` and `/lib64`, etc. aren't really reliable. That means that anything compiled in a way other than 110% statically is unlikely to run. I've even seen some binaries that _are_ statically-linked not work on NixOS.

Someone can correct me if I'm wrong but that's been my experience.

Source for the LD stuff not working: https://github.com/NixOS/nixpkgs/issues/107375#issuecomment-749580559
This holds true still.

**Comment by marcusramberg** [comment]
I was the one who opened the PR to bring it back, but having said that I'm not sure I love the cargo2nix stuff as it seems like you'd need to keep it in sync with the cargo file as aurae changes? The reason I opened it was the `flake.nix` file still in the repo was referring to it. Having said that I don't love the cargo2nix stuff as it seems like you'd have to keep this file up to date, and like Nova points out it adds a lot of auto generated code to the project. Isn't there a simpler rust builder we could use? I guess the intent when adding this was for actual development of Aurae and not distribution, however I found it easier to just use rustup inside Nixos as well.

**Comment by hbjydev** [comment]
Using Rust with Nix gets challenging as soon as you start to include system dependencies (a notoriously hard problem with Nix and Rust is crates that depend on OpenSSL, as setting OpenSSL's LD flags are kinda hard. Not an issue here, but it's just an example). So I feel like Nix in this is probably not all that useful, but it _does_ alienate developers who use NixOS on their systems somewhat, because their dev environments won't be particularly frictionless.

I'm not entirely aware of a better system than cargo2nix, but I do wonder if it could be made to work by just including `cargo` and `rustc` in the flake instead

**Comment by evanrichter** [comment]
I've used naersk to build cargo projects with rust, but even that isn't necessary if all you want is a reproducible dev environment.

You can probably just use flake.nix and it's relatively small auto generated flake.lock. and use oxalica/rust-overlay for the rust toolchain

**Comment by krisnova** [comment]
so while I appreciate the context, opinions, and collective experience of the above comments they aren't answering my question which of course is: 

Why is the Cargo.nix file here? 

Or more importantly: Why should it remain here? 

I merged the file in originally with @edude03 at KubeCon and I guess what I am getting at is that I'm not really seeing any strong cases for keeping it around. 

Maybe another way of framing the discussion: 

What will we break if we remove the Cargo.nix file and the associated flake things as well?

**Comment by hbjydev** [comment]
@krisnova I'm not sure it makes sense in there, as the `Cargo.nix` file isn't really package management as much as it is a lockfile, like you mentioned before, and there would have to be some weird CI to make that work well. Either way, I believe Cargo2nix works by reading the Cargo.lock and generating the .nix equivalent from it, so it may break Cargo2nix. As far as I can see it's not used in Aurae's CI, but I'm sure there would be people who want to run it in prod with Nix.

**Comment by evanrichter** [comment]
I will say, even though I haven't contributed to Aurae, that having a flake.nix/lock that some maintainers _do use_ is going to be super inviting for new contributors. Since I won't have to fuss with any development libraries, specific rust toolchains and targets, etc. I'll just have to run `nix develop` and be ready to go. That applies to any devs using nix, not just NixOS.

Cargo2nix seems like (just one) nix way to get reproducible rust builds. If your project isn't getting benefit from that, I would say remove the cargo.nix because it does skew the metrics visual.

But please don't get rid of all nix files. It really does make sense for flake.nix to live in-repo

**Comment by edude03** [comment]
Hey as the guy who originally added the nix stuff I feel like I should
chime in, though I’m on my phone so I’ll probably have to follow up -

So I agree with nova about project drift and using cargo2nix was the wrong
choice for this. My original plan was to include an update script but I
think the effort outweighs the benefit since someone still has to run the
update script and commit the updates.

My updated plan (which I don’t mind implementing and maintaining) is to
only use nix to provide the devshell, so rustc cargo and protoc.

It can call the makefile however in that case we should also have CI to
make sure it still works.

**Comment by krisnova** [comment]
Thanks for responding @edude03. Your proposal sounds good to me! Thank you for stepping up and your leadership here. Obviously I will trust whatever you think is best! ❤️ 

I think all of my concerns are addressed as long as we aren't setting a false expectation that `nix` will somehow build the software different than the rest of the project I am okay with having the project support Nix 😄 

Feel free to close this issue once your suggested changes are merged into main!

Thanks everyone!

**Comment by jhillyerd** [comment]
@edude03 please take a look at #448 -- I did the minimum to get `make auraed-build` and `make auraescript-build`.

As noted in the PR, Makefile expects a mutable rustup environment, when I just tried to provide the rust-overlay stable toolchain directly, I got:

```
$ make auraed-build                                       
rustup target list | grep -qc 'x86_64-unknown-linux-musl (installed)' || \                                                                                                 
rustup target add x86_64-unknown-linux-musl                                                                                                                                
error: no override and no default toolchain set                                                                                                                            
error: no override and no default toolchain set                                      
make: *** [Makefile:134: musl] Error 1    
```

So I added rustup and the shell hook to set a default toolchain instead.

**Comment by jhillyerd** [comment]
#464 should close this issue when merged.  This could be revisited when Aurae has some stable releases; if it makes sense to run auraed or auraescript on-top of NixOS, a separate repo could be created to contain the flake and lock files.  That flake repo could be updated with each release.
